vllm/engine/async_llm_engine.py

```diff
@@ -296,6 +360,66 @@ async def step_async(
 
         return request_outputs
 
+    def _has_remaining_steps(
+        self, seq_group_metadata_list: Optional[List[SequenceGroupMetadata]]
+    ) -> bool:
+        if not self.scheduler_config.is_multi_step:
+            return False
+
+        if seq_group_metadata_list is None:
+            return False
+        # TODO(will) this is a sanity check for nowto make sure that all the
+        # seqs are on the same steps. Eventually we will want to do some sort of
+        # dynamic scheduling when doing multi-step decoding.
+        if len(seq_group_metadata_list) == 0:
+            return False
+        steps_remaining = [
+            seq_group.state.remaining_steps
+            for seq_group in seq_group_metadata_list
+        ]
+        if steps_remaining.count(steps_remaining[0]) != len(steps_remaining):
+            raise AssertionError(("All running sequence groups should "
+                                  "have the same remaining steps."))
+
+        if any(seq_group.state.remaining_steps > 0
+               for seq_group in seq_group_metadata_list):
+            return True
+        return False
+
+    def _cache_scheduler_outputs_for_multi_step(
+            self, virtual_engine: int,
+            seq_group_metadata_list: Optional[List[SequenceGroupMetadata]],
+            scheduler_outputs: SchedulerOutputs) -> None:
+        self.cached_scheduler_outputs[
+            virtual_engine].seq_group_metadata_list = seq_group_metadata_list
+        self.cached_scheduler_outputs[virtual_engine].scheduler_outputs = \
+            scheduler_outputs
+        self.cached_scheduler_outputs[virtual_engine].last_output = None
+
+    def _get_cached_sampled_token_ids_for_multi_step(
+            self, virtual_engine: int) -> Optional[torch.Tensor]:
+        cached_last_output = self.cached_scheduler_outputs[
+            virtual_engine].last_output
+        if (self.scheduler_config.is_multi_step
+                and self.parallel_config.pipeline_parallel_size > 1
+                and cached_last_output is not None
+                and cached_last_output.sampled_token_ids_numpy is not None):
+            return torch.from_numpy(cached_last_output.sampled_token_ids_numpy)
+        return None
+
+    def _cache_output_for_multi_step(
+            self, virtual_engine: int,
+            output: List[Optional[SamplerOutput]]) -> None:
+        if (self.parallel_config.pipeline_parallel_size > 1 and len(output) > 0
+                and output[0] is not None):
+            last_output = output[-1]
+            assert last_output is not None
+            assert last_output.sampled_token_ids_numpy is not None
+            assert last_output.sampled_token_ids is None
+            assert last_output.sampled_token_probs is None
+            self.cached_scheduler_outputs[
+                virtual_engine].last_output = last_output
```

Comment:
Are these functions merely for PP? Or are they required for multi-step in general?

Reply:
They are only needed for PP. Used to cache and pass the sampled_token_ids tensor from the previous step to all of the non-last rank PP stages in order for them to perform in-place prepare_input

Reply:
For TP, we use the existing input_preparation broadcast to communicate this tensor